Web Driver: navigate endpoint needs to correctly handle state and url validation

AnujPanta1 opened the pull request:
#### b7ee2350d379892cf427dd0b7dcf7b66c762012d
<pre>
Web Driver: navigate endpoint needs to correctly handle state and url validation
<a href="https://bugs.webkit.org/show_bug.cgi?id=296634">https://bugs.webkit.org/show_bug.cgi?id=296634</a>
<a href="https://rdar.apple.com/157031091">rdar://157031091</a>

Reviewed by BJ Burg.

Validation of the URL parameter against the browsing-context’s URL.
Also, changing the default readiness state to match the spec.

* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::navigate):

Canonical link: <a href="https://commits.webkit.org/299780@main">https://commits.webkit.org/299780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f15a9014d5e79298144519a45c64be2ff1e46bae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30443 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126442 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72171 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48369 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91208 "21 flakes 78 failures") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60513 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32342 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107680 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71759 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31378 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70070 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101821 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129351 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47018 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99824 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47384 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99667 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25315 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45122 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23142 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43649 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46880 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52586 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46348 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49695 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48032 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c8a086724323edd327dc94cec9fc2dda4cc3efc0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48660

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
```diff
@@ -291,15 +292,29 @@ void BidiBrowsingContextAgent::navigate(const BrowsingContext& browsingContext,
 
-    auto pageLoadStrategy = pageLoadStrategyFromReadinessState(optionalReadinessState.value_or(defaultReadinessState));
-    session->navigateBrowsingContext(browsingContext, url, pageLoadStrategy, defaultPageLoadTimeout.milliseconds(), [url, callback = WTFMove(callback)](CommandResult<void>&& result) {
+    RefPtr webPageProxy = session->webPageProxyForHandle(browsingContext);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!webPageProxy, WindowNotFound);
+
+    auto waitCondition = optionalReadinessState.value_or(defaultReadinessState);
+    if (optionalReadinessState && *optionalReadinessState == ReadinessState::None)
+        waitCondition = ReadinessState::None;
+    else if (!optionalReadinessState)
+        waitCondition = ReadinessState::Interactive;
+
+    URL baseURL = URL(webPageProxy->currentURL());
+    URL urlRecord(baseURL, url);
+
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!urlRecord.isValid(), InvalidParameter);
+
+    auto pageLoadStrategy = pageLoadStrategyFromReadinessState(waitCondition);
```
Doesn't the `value_or` already achieve what you want from the `if` and `else if`?

```suggestion
    RefPtr webPageProxy = session->webPageProxyForHandle(browsingContext);
    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!webPageProxy, WindowNotFound);

    URL baseURL(webPageProxy->currentURL());
    URL urlRecord(baseURL, url);
    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!urlRecord.isValid(), InvalidParameter);

    auto pageLoadStrategy = pageLoadStrategyFromReadinessState(optionalReadinessState.value_or(defaultReadinessState));
```

AnujPanta1 replied:
Yes you are right, I think when I merged my branches I left this if statement !

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
```diff
@@ -291,15 +292,29 @@ void BidiBrowsingContextAgent::navigate(const BrowsingContext& browsingContext,
+    else if (!optionalReadinessState)
+        waitCondition = ReadinessState::Interactive;
+
+    URL baseURL = URL(webPageProxy->currentURL());
```
We prefer using brace initializers where possible:

URL baseURL { webPageProxy->currentURL()) };
URL urlRecord { baseURL, url };

AnujPanta1 committed:
Web Driver: navigate endpoint needs to correctly handle state and url validation
https://bugs.webkit.org/show_bug.cgi?id=296634
rdar://157031091

Reviewed by BJ Burg.

Validation of the URL parameter against the browsing-context’s URL.
Also, changing the default readiness state to match the spec.

* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::navigate):

Canonical link: https://commits.webkit.org/299780@main
